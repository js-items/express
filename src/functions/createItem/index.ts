import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { CREATED } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import sendResponse from '../../utils/sendResponse';

const createItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeCreateItem
  );

  await transactionHandler({ req, res }, async () => {
    const { item } = await config.service.createItem({
      id: req.body.id,
      item: config.convertDocumentIntoItem({ document: req.body, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item, req, res }),
      status: CREATED,
    });
  });
};

export default createItem;
