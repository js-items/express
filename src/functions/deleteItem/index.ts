import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { NO_CONTENT } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const deleteItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeDeleteItem
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');
    
    await config.service.deleteItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
    });

    sendResponse({
      config,
      req,
      res,
      status: NO_CONTENT,
    });
  });
};

export default deleteItem;
