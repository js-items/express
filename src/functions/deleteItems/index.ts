import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { NO_CONTENT } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const deleteItems: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeDeleteItems
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');
    
    await config.service.deleteItems({
      filter: config.createFilter({ filter, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      status: NO_CONTENT,
    });
  });
};

export default deleteItems;
