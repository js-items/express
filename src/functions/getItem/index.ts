import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const getItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeGetItem
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const { item } = await config.service.getItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
    });
    
    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item, req, res }),
      status: OK,
    });
  });
};

export default getItem;
