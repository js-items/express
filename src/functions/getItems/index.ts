import { Item } from '@js-items/foundation';
import boolean from 'boolean';
import { Request, Response } from 'express';
import { OutgoingHttpHeaders } from 'http';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import _isNil from 'ramda/src/isNil';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import getNumberQueryParam from '../../utils/getNumberQueryParam';
import sendResponse from '../../utils/sendResponse';

const getItems: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeGetItems
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const sort = !_isNil(req.query.sort)
      ? getJsonQueryParam(req.query, 'sort')
      : config.defaultSort;
      
    const limit = getNumberQueryParam(
      req.query,
      'limit',
      config.defaultPaginationLimit
    );
    
    const createdFilter = config.createFilter({ filter, req, res });

    const { cursor, items } = await config.service.getItems({
      filter: createdFilter,
      pagination: {
        after: req.query.after,
        before: req.query.before,
        limit,
      },
      sort,
    });

    const { count: totalCount } = await config.service.countItems({
      filter: createdFilter,
    });

    const responseHeaders: OutgoingHttpHeaders = {
      [config.afterHeaderName]: cursor.after,
      [config.beforeHeaderName]: cursor.before,
      [config.hasBeforeHeaderName]: cursor.hasBefore.toString(),
      [config.hasAfterHeaderName]: cursor.hasAfter.toString(),
      [config.totalHeaderName]: totalCount,
    };

    const responseData = items.map(item =>
      config.convertItemIntoDocument({ item, req, res })
    );

    const nestedObject = {
      [config.paginationKey]: {
        [config.afterKey]: cursor.after,
        [config.beforeKey]: cursor.before,
        [config.hasBeforeKey]: cursor.hasBefore,
        [config.hasAfterKey]: cursor.hasAfter,
        [config.totalKey]: totalCount,
      },
      [config.dataKeyName]: responseData,
    };

    const enveloped = boolean(req.query[config.envelopeParamName]);

    const responseObject = enveloped ? nestedObject : responseData;
    const headers = enveloped ? {} : responseHeaders;

    sendResponse({ req, res, config, status: OK, headers, responseObject });
  });
};

export default getItems;
