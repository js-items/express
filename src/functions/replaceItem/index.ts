import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const replaceItem: RequestHandlerFactory =  <I extends Item>(config: FacadeConfig<I>) => async (
  req: Request,
  res: Response
) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(config.beforeReplaceItem);

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const result = await config.service.replaceItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
      item: config.convertDocumentIntoItem({ document: req.body, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item: result.item, req, res }),
      status: OK,
    });
  });
};

export default replaceItem;