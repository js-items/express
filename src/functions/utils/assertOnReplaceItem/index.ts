// tslint:disable:no-any
import { ReplaceItem } from '@js-items/foundation';
import { TestItem } from '@js-items/foundation/dist/functions/utils/testItem';
import { OK } from 'http-status-codes';
import supertest from 'supertest';
import createTestUrl from '../createTestUrl';

const assertOnReplaceItem = async <I extends TestItem>({
  client,
  status = OK,
  params,
  body,
  expectedParams,
  url,
  replaceItem,
}: Options<I>) => {
  const requestUrl =
    params !== undefined ? createTestUrl({ url, params }) : url;

  const result = await client.put(requestUrl).send(body);

  expect(result.status).toBe(status);
  expect(result.body).toMatchSnapshot();

  if (expectedParams !== undefined) {
    expect(replaceItem).toBeCalledWith(expectedParams);
  }
};

export interface Options<I extends TestItem> {
  readonly client: supertest.SuperTest<supertest.Test>;
  readonly status?: number;
  readonly params?: any;
  readonly body?: any;
  readonly expectedParams?: { item: any; id: string };
  readonly url: string;
  readonly replaceItem: ReplaceItem<I>;
}

export default assertOnReplaceItem;
