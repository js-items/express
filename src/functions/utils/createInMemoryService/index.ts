// tslint:disable:no-object-literal-type-assertion
import { Item } from '@js-items/foundation';
import Facade from '@js-items/foundation/dist/Facade';
import {
  Options as CreateItemOptions,
  Result as CreateItemResult,
} from '@js-items/foundation/dist/functions/CreateItem';
import {
  Options as GetItemOptions,
  Result as GetItemResult,
} from '@js-items/foundation/dist/functions/GetItem';
import { Result as GetItemsResult } from '@js-items/foundation/dist/functions/GetItems';
import { Result as ReplaceItemResult } from '@js-items/foundation/dist/functions/ReplaceItem';
import { Result as UpdateItemResult } from '@js-items/foundation/dist/functions/UpdateItem';
import { start } from '@js-items/foundation/dist/interfaces/Cursor';
import FactoryConfig from './FactoryConfig';

const defaultPromise = () => Promise.resolve();
/* istanbul ignore next */
const createInMemoryService = <I extends Item>(
  factoryConfig: FactoryConfig<I>
): Facade<I> => ({
  countItems: () => Promise.resolve({ count: 1 }),
  createItem: ({ item }: CreateItemOptions<I>) =>
    Promise.resolve({ item } as CreateItemResult<I>),
  deleteItem: defaultPromise,
  deleteItems: defaultPromise,
  getItem: ({ id }: GetItemOptions<I>) =>
    Promise.resolve({ item: { id } } as GetItemResult<I>),
  getItems: () =>
    Promise.resolve({
      cursor: {
        after: start,
        before: start,
        hasAfter: false,
        hasBefore: false,
      },
      items: [{ id: '1' }, { id: '2' }],
    } as GetItemsResult<I>),
  replaceItem: () =>
    Promise.resolve({ item: { id: '2' } } as ReplaceItemResult<I>),
  updateItem: () =>
    Promise.resolve({ item: { id: '1' } } as UpdateItemResult<I>),
  ...factoryConfig,
});

export default createInMemoryService;
