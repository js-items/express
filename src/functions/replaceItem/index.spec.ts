// tslint:disable:no-any
import * as sourceMapSupport from 'source-map-support';
import assertOnReplaceItem from '../utils/assertOnReplaceItem';
import initTests from '../utils/initTests';
sourceMapSupport.install();
import dotenv from 'dotenv';
dotenv.config();
import { ItemNotFoundError } from '@js-items/foundation';
import { Options as ReplaceItemOptions } from '@js-items/foundation/dist/functions/ReplaceItem';
import testItem, {
  TestItem,
} from '@js-items/foundation/dist/functions/utils/testItem';
import { NOT_FOUND, OK } from 'http-status-codes';
import { TEST_URL } from '../../constants';
import createInMemoryService from '../utils/createInMemoryService';

jest.mock('uuid', () => ({
  v4: jest.fn(() => '1'),
}));

const notExistingId = 'not-existing-id';

const replaceItem = jest.fn(
  async ({ item, id }: ReplaceItemOptions<TestItem>) => {
    if (id !== testItem.id) {
      return Promise.reject(new ItemNotFoundError('item', testItem.id));
    }

    return Promise.resolve({ item });
  }
);

describe('@replaceItem', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
  });

  const service = createInMemoryService({
    replaceItem,
  });

  const { request } = initTests<TestItem>({ service });

  const expectedParams = {
    filter: {},
    id: testItem.id,
    item: testItem,
  };

  const defaultOptions = {
    body: testItem,
    client: request,
    expectedParams,
    replaceItem,
    url: `${TEST_URL}/${testItem.id}`,
  };

  const notFoundOptions = {
    body: { ...testItem, id: notExistingId },
    expectedParams: {
      ...expectedParams,
      id: notExistingId,
      item: { ...testItem, id: notExistingId },
    },
  };

  it('throws not found error when item does exist', async () => {
    await assertOnReplaceItem({
      ...defaultOptions,
      ...notFoundOptions,
      status: NOT_FOUND,
      url: `${TEST_URL}/${notExistingId}`,
    });
  });

  it('throws not found error when item does exist and envelope is enabled', async () => {
    await assertOnReplaceItem({
      ...defaultOptions,
      ...notFoundOptions,
      params: {
        envelope: true,
      },
      status: OK,
      url: `${TEST_URL}/${notExistingId}`,
    });
  });

  it('replaces item', async () => {
    const updatedItem = {
      booleanProperty: true,
      id: testItem.id,
      numberProperty: 999,
      stringProperty: 'testowo',
    };

    await assertOnReplaceItem({
      ...defaultOptions,
      body: updatedItem,
      expectedParams: {
        ...expectedParams,
        item: updatedItem,
      },
    });
  });
  // tslint:disable-next-line:max-file-line-count
});
