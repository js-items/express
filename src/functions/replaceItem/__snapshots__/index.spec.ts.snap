// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@replaceItem replaces item 1`] = `
"{
  \\"booleanProperty\\": true,
  \\"id\\": \\"6da9dce6-be40-4fac-bd9c-6a2f348a1c87\\",
  \\"numberProperty\\": 999,
  \\"stringProperty\\": \\"testowo\\"
}"
`;

exports[`@replaceItem throws not found error when item does exist 1`] = `
"{
  \\"itemId\\": \\"6da9dce6-be40-4fac-bd9c-6a2f348a1c87\\",
  \\"itemName\\": \\"item\\",
  \\"transactionId\\": \\"1\\"
}"
`;

exports[`@replaceItem throws not found error when item does exist and envelope is enabled 1`] = `
"{
  \\"headers\\": {
    \\"x-powered-by\\": \\"Express\\"
  },
  \\"response\\": {
    \\"itemId\\": \\"6da9dce6-be40-4fac-bd9c-6a2f348a1c87\\",
    \\"itemName\\": \\"item\\",
    \\"transactionId\\": \\"1\\"
  },
  \\"status\\": 404
}"
`;
